### PR TITLE
Make thermal compensation work with legacy calibrators

### PR DESCRIPTION
**Pull request by priseborough**
This has been submitted as a placeholder to enable feedback and assistance with integration issues.

The objective was to enable the commander controlled gyro and accel calibration to be used, whilst preserving the integrity and usefulness of exisiting thermal compensation. The objective is to have a workflow where thermal calibration can be done on a new board first, followed by the normal gyro and accel calibration when installed in a vehicle.

Testing on a pixracer has shown that the thermal calibration parameters for the second accel are not being retrieved correctly.

Changes have been made to try and match the sensor ID in the parameters to the sensor ID in the uORB when setting mappings.

There is hack to the sensor drivers so that the scale factors and offsets for the IMU accels can be zero, because when we are using thermal compensation, the accel calibrator needs to work with the thermal compensation TC_A* parameters, not the CAL_* parameters used by the drivers. 

@jgoppert and @bugobliterator note the change to sign convention for offsets and the change to order of application of offset and scale factor. We now use the convention consistent with the legacy accel calibrator that corrected_value = (raw_value - offset) *scale_factor

### CODE REVIEW

**Comment by LorenzMeier** [comment]
@bugobliterator Where are you standing with this?

**Comment by LorenzMeier** [comment]
Sounds like it needs squashing.

File: src/modules/mc_att_control/mc_att_control_main.cpp [hunk 1]
```diff
@@ -994,9 +993,28 @@ MulticopterAttitudeControl::control_attitude_rates(float dt)
 	_board_rotation = board_rotation_offset * _board_rotation;
 
 	// get the raw gyro data and correct for thermal errors
-	math::Vector<3> rates(_sensor_gyro.x * _sensor_correction.gyro_scale[0] + _sensor_correction.gyro_offset[0],
-			      _sensor_gyro.y * _sensor_correction.gyro_scale[1] + _sensor_correction.gyro_offset[1],
-			      _sensor_gyro.z * _sensor_correction.gyro_scale[2] + _sensor_correction.gyro_offset[2]);
+	math::Vector<3> rates;
+
+	if (_sensor_correction.gyro_mapping[_selected_gyro] == 0) {
```
Comment (bkueng) [inline src/modules/mc_att_control/mc_att_control_main.cpp]: @priseborough I don't think this is correct: _selected_gyro is the uorb instance and gyro_offset_0 contains data from uorb instance 0, meaning the mapping is not needed.

**Comment by bkueng** [comment]
We can squash & cleanup the commits when rebasing. I wanted to make sure my changes can be followed easily.

**Comment by LorenzMeier** [comment]
@bkueng Can you finish the fix up where needed and apply it to master?

File: src/modules/mc_att_control/mc_att_control_main.cpp (hunk 1 above)
Comment (priseborough) [inline src/modules/mc_att_control/mc_att_control_main.cpp]: gyro_offset_0 contains the  data from the uORB 0 instance when the cal was performed, but I did not think there was a guarantee that the same sensor ID would always appear on the same uORB instance, hence why I was intending to match device ID when setting the mapping from uORB instance to parameter index.

Comment (LorenzMeier) [inline src/modules/mc_att_control/mc_att_control_main.cpp]: That's correct - there is no guarantee at all that this matches between calibration time and runtime.

**Comment by priseborough** [comment]
The accel thermal correction parameters are now being retrieved by the apply_corrections_accel function. I will now track down what is causing a sign error in the thermal offset parameter adjustment made by the accel calibrator. The sing convention used by the commander accel calibrator is not explicitly documented, so i need to go through the code in more detail to work out where the error is coming from rather than just changing signs until it works.

**Comment by bkueng** [comment]
I rebased on master so it includes #6425, and we can further test with this branch.

So far I verified that normal calibration w/o temp comp still works, calibration with temp comp enabled for gyro & accel work too. But I used default values for temp comp, and I could not reproduce the sign error from @priseborough.

Also with temp comp enabled, the gyro selftests fail.

@priseborough there was a merge conflict with `process_sensor_caldata.py` (commit 'Tools: change definition of sensor offset and scale factor to match legacy code'). I chose the version from master, please verify this is correct.

Next step is to actually execute a temp calibration process.

**Comment by LorenzMeier** [comment]
Jenkins test this please.

**Comment by priseborough** [comment]
@bkueng  and @jgoppert  I have adjusted the sign convention used by the version of process_sensor_caldata.py to match the flight code, however in the process of doing this I have found that there are bugs in the script that casues it to not print all of the  required values to the .param file and also not allow for the effect of gravitational offset on the accel readings.

Until these issues are fixed, it would be better to use the old version of the calibrator and not the one in master.

**Comment by bkueng** [comment]
I tested the whole temperature calibration several times, with the onboard `send_event temperature_calibration` (gyro only) method, while logging the data at the same time, on a non-problematic (see #6449) pixracer, only 1 sensor.

The resulting parameters match with the ones generated from process_sensor_caldata.py script (old version). @bugobliterator I had to reverse the order of the polynom coefficients, see https://github.com/PX4/Firmware/pull/6419/commits/454279031d3ef00e55e3df6ba9bfeea60f883eff


As a comparison, onboard generated params (before reordering):
```
1	1	TC_G0_ID	2360328	6
1	1	TC_G0_SCL_0	1.000000000000000000	9
1	1	TC_G0_SCL_1	1.000000000000000000	9
1	1	TC_G0_SCL_2	1.000000000000000000	9
1	1	TC_G0_TMAX	56.157894134521484375	9
1	1	TC_G0_TMIN	32.155124664306640625	9
1	1	TC_G0_TREF	44.155124664306640625	9
1	1	TC_G0_X0_0	0.000000531149453309	9
1	1	TC_G0_X0_1	-0.000000053815089274	9
1	1	TC_G0_X0_2	-0.000000010459545230	9
1	1	TC_G0_X1_0	0.000000091002959834	9
1	1	TC_G0_X1_1	0.000000488552757361	9
1	1	TC_G0_X1_2	-0.000001751046283971	9
1	1	TC_G0_X2_0	0.000067804379796144	9
1	1	TC_G0_X2_1	-0.000148861756315455	9
1	1	TC_G0_X2_2	0.000114951129944529	9
1	1	TC_G0_X3_0	-0.020944913849234581	9
1	1	TC_G0_X3_1	0.008191471919417381	9
1	1	TC_G0_X3_2	-0.000357050244929269	9
```

and process_sensor_caldata.py:
```
1	1	TC_G0_ID	2360328	6
1	1	TC_G0_SCL_0	1.0	9
1	1	TC_G0_SCL_1	1.0	9
1	1	TC_G0_SCL_2	1.0	9
1	1	TC_G0_TMAX	56.9474	9
1	1	TC_G0_TMIN	31.1136	9
1	1	TC_G0_TREF	44.0304718018	9
1	1	TC_G0_X0_0	-0.0209974212402	9
1	1	TC_G0_X0_1	0.00818708289477	9
1	1	TC_G0_X0_2	-0.000372584207174	9
1	1	TC_G0_X1_0	6.97038819388e-05	9
1	1	TC_G0_X1_1	-0.000142561434139	9
1	1	TC_G0_X1_2	0.000111114656513	9
1	1	TC_G0_X2_0	2.75543609706e-07	9
1	1	TC_G0_X2_1	8.79902435918e-07	9
1	1	TC_G0_X2_2	-2.01106606351e-06	9
1	1	TC_G0_X3_0	4.6223905129e-07	9
1	1	TC_G0_X3_1	-1.29665377912e-07	9
1	1	TC_G0_X3_2	5.86254270456e-08	9
```
Here's the log: http://logs.px4.io/download?log=e948aba1-be80-432e-9fd4-04b5caa8a01d

**Comment by priseborough** [comment]
I as able to get the gyro and accel calibrator working. I will now start looking at why the baro thermal correction is a NaN.

**Comment by priseborough** [comment]
I got it working on the pixracer. I had to revert Tools/process_sensor_caldata.py back to a version before the refactoring to unblock my testing, but other than that it appears to work and I have followed the procedure documented in Tools/process_sensor_caldata.py and flight tested successfully.

Here is the calibration log used http://logs.px4.io/plot_app?log=7a3f865f-2013-472b-8345-cee7f3ec1e2d

**Comment by LorenzMeier** [comment]
@priseborough The board reboots when disconnecting USB, so most people will be fine. But we should enforce it.

**Comment by LorenzMeier** [comment]
Looks good to me. We should be able to bring this in.

**Comment by LorenzMeier** [comment]
Rebased and continued here: https://github.com/PX4/Firmware/pull/6471

**Pull request closed without merging**